core/drivers/clk/clk-stm32mp25.c

```
@@ -0,0 +1,3503 @@
+	uint32_t fracin = 0;
+	uint32_t fbdiv = 0;
+	uint32_t refdiv = 0;
+
+	fracin = io_read32(pllxcfgr3) & RCC_PLLxCFGR3_FRACIN_MASK;
+	fbdiv = (io_read32(pllxcfgr2) & RCC_PLLxCFGR2_FBDIV_MASK) >>
+		RCC_PLLxCFGR2_FBDIV_SHIFT;
+
+	refdiv = io_read32(pllxcfgr2) & RCC_PLLxCFGR2_FREFDIV_MASK;
+
+	if (fracin != 0U) {
+		unsigned long long numerator, denominator;
+
+		numerator = ((unsigned long long)fbdiv << 24) + fracin;
+		numerator = prate * numerator;
+		denominator = (unsigned long long)refdiv << 24;
+		fvco = (unsigned long)(numerator / denominator);
+	} else {
+		fvco = (unsigned long)(prate * fbdiv / refdiv);
+	}
+
+	return fvco;
+}
+
+static unsigned long clk_stm32_pll_get_rate(struct clk *clk __unused,
+					    unsigned long prate)
+{
+	struct clk_stm32_priv *priv = clk_stm32_get_priv();
+	struct clk_stm32_pll_cfg *cfg = clk->priv;
+	uintptr_t pllxcfgr1 = priv->base + cfg->pll_offset;
+	uintptr_t pllxcfgr4 = pllxcfgr1 + RCC_OFFSET_PLLXCFGR4;
+	uintptr_t pllxcfgr6 = pllxcfgr1 + RCC_OFFSET_PLLXCFGR6;
+	uintptr_t pllxcfgr7 = pllxcfgr1 + RCC_OFFSET_PLLXCFGR7;
+	unsigned long dfout = 0;
+	uint32_t postdiv1 = 0;
+	uint32_t postdiv2 = 0;
+
+	postdiv1 = io_read32(pllxcfgr6) & RCC_PLLxCFGR6_POSTDIV1_MASK;
+	postdiv2 = io_read32(pllxcfgr7) & RCC_PLLxCFGR7_POSTDIV2_MASK;
+
+	if ((io_read32(pllxcfgr4) & RCC_PLLxCFGR4_BYPASS) != 0U) {
+		dfout = prate;
+	} else {
+		if (postdiv1 == 0U || postdiv2 == 0U)
+			dfout = prate;
+		else
+			dfout = clk_get_pll_fvco(cfg->pll_offset, prate) /
+						 (postdiv1 * postdiv2);
+	}
+
+	return dfout;
+}
+
+static TEE_Result clk_stm32_pll_enable(struct clk *clk)
+{
+	struct clk_stm32_pll_cfg *cfg = clk->priv;
+
+	if (stm32_gate_rdy_enable(cfg->gate_id) != 0U)
+		EMSG("%s timeout", clk_get_name(clk));
+
+	return TEE_SUCCESS;
+}
+
+static void clk_stm32_pll_disable(struct clk *clk)
+{
+	struct clk_stm32_pll_cfg *cfg = clk->priv;
+
+	if (stm32_gate_rdy_disable(cfg->gate_id) != 0U)
+		EMSG("%s timeout", clk_get_name(clk));
```

is it okay to silently fail on a timeout issue?